cmd/clusterctl/client/tree/discovery.go

```diff
@@ -105,25 +119,112 @@ func Discovery(ctx context.Context, c client.Client, namespace, name string, opt
 	controlPlaneMachines := selectControlPlaneMachines(machinesList)
 	for i := range controlPlaneMachines {
 		cp := controlPlaneMachines[i]
-		addMachineFunc(controlPLane, cp)
+		addMachineFunc(controlPlane, cp)
 	}
 
-	if len(machinesList.Items) == len(controlPlaneMachines) {
+	machinePoolList, err := getMachinePoolsInCluster(ctx, c, cluster.Namespace, cluster.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(machinesList.Items) == len(controlPlaneMachines) && len(machinePoolList.Items) == 0 {
 		return tree, nil
 	}
 
 	workers := VirtualObject(cluster.Namespace, "WorkerGroup", "Workers")
 	tree.Add(cluster, workers)
 
+	err = addMachineDeploymentToObjectTree(ctx, c, cluster, workers, machinesList, tree, options, addMachineFunc)
+	if err != nil {
+		return nil, err
+	}
+
+	addMachinePoolsToObjectTree(cluster, workers, machinePoolList, tree, options)
+
+	// Handles orphan machines.
+	if len(machineMap) < len(machinesList.Items) {
+		other := VirtualObject(cluster.Namespace, "OtherGroup", "Other")
+		tree.Add(workers, other)
+
+		for i := range machinesList.Items {
+			m := &machinesList.Items[i]
+			if _, ok := machineMap[m.Name]; ok {
+				continue
+			}
+			addMachineFunc(other, m)
+		}
+	}
+
+	return tree, nil
+}
+
+func constructPlaceholderObject(objectRef *corev1.ObjectReference) *unstructured.Unstructured {
+	placeholder := unstructured.Unstructured{}
+	placeholder.SetKind(objectRef.Kind)
+	placeholder.SetNamespace(objectRef.Namespace)
+	placeholder.SetName(objectRef.Name)
+	placeholder.SetAPIVersion(objectRef.APIVersion)
+	placeholder.SetUID(types.UID(objectRef.Kind + "/" + objectRef.Namespace + "/" + objectRef.Name))
+
+	return &placeholder
+}
+
+func addClusterResourceSetsToObjectTree(ctx context.Context, c client.Client, cluster *clusterv1.Cluster, tree *ObjectTree) {
+	if resourceSetBinding, err := getResourceSetBindingInCluster(ctx, c, cluster.Namespace, cluster.Name); err == nil {
+		resourceSetGroup := VirtualObject(cluster.Namespace, "ClusterResourceSetGroup", "ClusterResourceSets")
+		tree.Add(cluster, resourceSetGroup)
+
+		owners := make(map[string]struct{})
+		for _, owner := range resourceSetBinding.GetOwnerReferences() {
+			if owner.Kind == "ClusterResourceSet" {
+				owners[owner.Name] = struct{}{}
+			}
+		}
+		for _, binding := range resourceSetBinding.Spec.Bindings {
+			resourceSetPlaceholder := constructPlaceholderObject(&corev1.ObjectReference{
```

Comment:
Is there a reason why we are not using a Virtual object here? Otherwise, let's create `ObjectReferenceObject` instead of `placeHolderObject` which is a little bit too generic

Reply:
I was under the impression that VirtualObjects were used to represent "dummy nodes" we added that don't correspond to an underlying CRD. The VirtualObject constructor overwrites the apiVersion as `virtual.cluster.x-k8s.io/v1beta1` which loses the group and version info in the CRD, which I end up using to color code the nodes.

Reply:
let's rename it `ObjectReferenceObject`, placeHolderObject is a little bit too generic